docs/adr-custom-exceptions.md

```diff
@@ -0,0 +1,23 @@
+# InstructLab Custom Exceptions Pattern
+
+## Context
+
+The historically pervasive pattern for exception handling during application runtime is to catch internally raised exceptions in the CLI layer and use `click.secho` followed by `click.exceptions.Exit` to display a useful error message to the user before exiting the application. This leaves a risk of intermediate calls between the site of the exception and the user-facing layer changing, leading to missed new exceptions and outdated caught exceptions. A second issue is that of discoverability and verification: given a `click.exceptions.Exit` exception handling, it is not clear from the code where the caught exception originates from in the call stack, and, similarly, given a piece of code that can raise an exception, it is not clear from the local code whether that exception is properly handled in the CLI layer without investigation.
+
+These issues will compound whenever REST APIs begin development.
+
+## Decision
+
+InstructLab will adopt custom exceptions to handle specific faults which are then caught in the CLI layer.
```

Comment:
And separately caught in the REST layer I assume, since the handling would be different there.  For the SDK layer, I guess we want to just pass them along?  Maybe that decision is out of scope for this document.

Reply:
I would say they should be caught by whatever the "caller" of the logic is - whether it's the CLI, a REST API, another Python application importing `instructlab`, etc.